homeassistant/components/nws/__init__.py

```diff
@@ -58,36 +59,55 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     nws_data = SimpleNWS(latitude, longitude, api_key, client_session)
     await nws_data.set_station(station)
 
-    async def update_observation() -> None:
-        """Retrieve recent observations."""
-        await call_with_retry(
-            nws_data.update_observation,
-            RETRY_INTERVAL,
-            RETRY_STOP,
-            start_time=utcnow() - UPDATE_TIME_PERIOD,
-        )
-
-    async def update_forecast() -> None:
-        """Retrieve twice-daily forecsat."""
-        await call_with_retry(
-            nws_data.update_forecast,
-            RETRY_INTERVAL,
-            RETRY_STOP,
-        )
-
-    async def update_forecast_hourly() -> None:
-        """Retrieve hourly forecast."""
-        await call_with_retry(
-            nws_data.update_forecast_hourly,
-            RETRY_INTERVAL,
-            RETRY_STOP,
-        )
-
+    def async_setup_update_observation(
+        retry_interval: datetime.timedelta | float,
+        retry_stop: datetime.timedelta | float,
+    ) -> Callable[[], Awaitable[None]]:
+        async def update_observation() -> None:
+            """Retrieve recent observations."""
+            await call_with_retry(
+                nws_data.update_observation,
+                retry_interval,
+                retry_stop,
+                start_time=utcnow() - UPDATE_TIME_PERIOD,
+            )
+
+        return update_observation
+
+    def async_setup_update_forecast(
+        retry_interval: datetime.timedelta | float,
+        retry_stop: datetime.timedelta | float,
+    ) -> Callable[[], Awaitable[None]]:
+        async def update_forecast() -> None:
+            """Retrieve twice-daily forecsat."""
+            await call_with_retry(
+                nws_data.update_forecast,
+                retry_interval,
+                retry_stop,
+            )
+
+        return update_forecast
+
+    def async_setup_update_forecast_hourly(
+        retry_interval: datetime.timedelta | float,
+        retry_stop: datetime.timedelta | float,
+    ) -> Callable[[], Awaitable[None]]:
+        async def update_forecast_hourly() -> None:
+            """Retrieve hourly forecast."""
+            await call_with_retry(
+                nws_data.update_forecast_hourly,
+                retry_interval,
+                retry_stop,
+            )
+
+        return update_forecast_hourly
```

Comment:
It looks like the inner function values never change so they could be written as partials
```diff
diff --git a/homeassistant/components/nws/__init__.py b/homeassistant/components/nws/__init__.py
index a4cdebcbe5e..df8cb4c329c 100644
--- a/homeassistant/components/nws/__init__.py
+++ b/homeassistant/components/nws/__init__.py
@@ -5,6 +5,7 @@ from __future__ import annotations
 from collections.abc import Awaitable, Callable
 from dataclasses import dataclass
 import datetime
+from functools import partial
 import logging
 
 from pynws import SimpleNWS, call_with_retry
@@ -78,30 +79,24 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
         retry_interval: datetime.timedelta | float,
         retry_stop: datetime.timedelta | float,
     ) -> Callable[[], Awaitable[None]]:
-        async def update_forecast() -> None:
-            """Retrieve twice-daily forecsat."""
-            await call_with_retry(
+        return partial(
+            call_with_retry,
             nws_data.update_forecast,
             retry_interval,
             retry_stop,
         )
 
-        return update_forecast
-
     def async_setup_update_forecast_hourly(
         retry_interval: datetime.timedelta | float,
         retry_stop: datetime.timedelta | float,
     ) -> Callable[[], Awaitable[None]]:
-        async def update_forecast_hourly() -> None:
-            """Retrieve hourly forecast."""
-            await call_with_retry(
+        return partial(
+            call_with_retry,
             nws_data.update_forecast_hourly,
             retry_interval,
             retry_stop,
         )
 
-        return update_forecast_hourly
-
     # Don't use retries in setup
     coordinator_observation = TimestampDataUpdateCoordinator(
         hass,

```